From: =?utf-8?q?Guido_G=C3=BCnther?= <agx@sigxcpu.org>
Date: Tue, 24 Sep 2024 13:26:22 +0200
Subject: build: Version the library name instead of the soname
MIME-Version: 1.0
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

Now with a real consumer (phrog) let's rather version the API and use a
fixed soname. This unconfuses some packaging tools and also avoids file
conflicts between different API versions.

Signed-off-by: Guido Günther <agx@sigxcpu.org>
Part-of: <https://gitlab.gnome.org/World/Phosh/phosh/-/merge_requests/1534>
---
 .gitlab-ci.yml  |  3 ++-
 meson.build     |  4 ++--
 src/meson.build | 12 ++++++------
 3 files changed, 10 insertions(+), 9 deletions(-)

diff --git a/.gitlab-ci.yml b/.gitlab-ci.yml
index 85565b6..d00972e 100644
--- a/.gitlab-ci.yml
+++ b/.gitlab-ci.yml
@@ -55,6 +55,7 @@ variables:
                     upower-dev wayland-dev wayland-protocols ttf-dejavu evolution-data-server-dev evince-dev
                     libadwaita-dev json-glib-dev gnome-bluetooth-dev modemmanager-dev gmobile-dev
   RUST_BINDINGS_BRANCH: main
+  LIBPHOSH: libphosh-0.42
 
 .trixie_vars: &trixie_vars
   variables:
@@ -131,7 +132,7 @@ build:lib-debian-trixie:
     - meson setup ${BUILD_OPTS} _build
     - meson compile -C _build 
     - meson install -C _build 
-    - gcc -Wall -O2 $(pkgconf --cflags libphosh-0) .gitlab-ci/libphosh-test.c $(pkgconf --libs libphosh-0)
+    - "gcc -Wall -O2 $(pkgconf --cflags $LIBPHOSH) .gitlab-ci/libphosh-test.c $(pkgconf --libs $LIBPHOSH)"
   artifacts:
     paths:
       - _build
diff --git a/meson.build b/meson.build
index b976972..0ed16dc 100644
--- a/meson.build
+++ b/meson.build
@@ -6,9 +6,9 @@ project('phosh', 'c',
 )
 
 app_id = 'sm.puri.Phosh'
-lib_inc_subdir = 'libphosh-0'
 ver_parts = meson.project_version().split('.')
-soversion = '@0@_@1@'.format(ver_parts[0], ver_parts[1])
+libversion = '@0@.@1@'.format(ver_parts[0], ver_parts[1])
+lib_inc_subdir = 'libphosh-@0@'.format(libversion)
 
 prefix = get_option('prefix')
 bindir = join_paths(prefix, get_option('bindir'))
diff --git a/src/meson.build b/src/meson.build
index 08f9223..1812be9 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -453,12 +453,12 @@ phosh_tool_dep = declare_dependency(
   dependencies: phosh_deps)
 
 # A library containing everything
-phosh_lib = both_libraries('phosh',
+phosh_lib = both_libraries('phosh-@0@'.format(libversion),
   libphosh_private_headers,
   libphosh_sources,
   include_directories: phosh_lib_inc,
   dependencies: [phosh_tool_dep, phosh_deps],
-  soversion: soversion,
+  soversion: 0,
   install: bindings_lib,
 )
 
@@ -491,7 +491,7 @@ if enable_introspection
                 sources : phosh_gir_sources,
               nsversion : '0',
               namespace : 'Phosh',
-        export_packages : 'libphosh-0',
+        export_packages : 'libphosh-@0@'.format(libversion),
           symbol_prefix : 'phosh',
       identifier_prefix : 'Phosh',
               link_with : phosh_lib,
@@ -507,10 +507,10 @@ if bindings_lib
   pkgconfig.generate(
     phosh_lib,
     name: 'libphosh',
-    subdirs: 'libphosh-0',
-    filebase: 'libphosh-0',
+    subdirs: 'libphosh-@0@'.format(libversion),
+    filebase: 'libphosh-@0@'.format(libversion),
     version : meson.project_version(),
-    description: 'Private library for building external components',
+    description: 'Library for building components embedding Phosh',
     requires: 'phosh-settings',
     url: 'https://world.pages.gitlab.gnome.org/Phosh/phosh/',
   )
